src/haddock/libs/libplots.py

```diff
@@ -523,29 +561,317 @@ def scatter_plot_handler(capri_filename, cl_rank, png, dpi):
+            automargin=True,
+            )
+        legend_title_text = sub_fig.layout.legend.title.text
+    fig.update_layout(
+        title_text=plot_title,
+        legend_title_text=legend_title_text,
+        height=height * number_of_rows,
+        width=width * number_of_cols,
+        )
+    return fig
+
+
+def find_best_struct(ss_file, number_of_struct=4):
+    """
+    Find best structures.
+
+    It inspects model-cluster-ranking recorded in capri_ss.tsv file and finds
+    the best models (models with lower ranks). By default, it selects the 4 best
+    models.
+
+    Parameters
+    ----------
+    ss_file : path
+        path to capri_ss.tsv
+    number_of_struct: int
+        number of models with lower model-cluster-ranking
+
+    Returns
+    -------
+    best_struct_df : pandas DataFrame
+        DataFrame of best structures
+    """
+    dfss = read_capri_table(ss_file)
+    dfss = dfss.sort_values(by=["cluster-id", "model-cluster-ranking"])
+    # TODO need a check for "Unclustered"
+    # possible number of structures, can be different per each cluster
+    max_number_of_struct = dfss.groupby("cluster-id").count()["model-cluster-ranking"].min()
+    number_of_struct = min(number_of_struct, max_number_of_struct)
+    best_struct_df = dfss.groupby("cluster-id").head(number_of_struct).copy()
+    number_of_cluster = len(best_struct_df["cluster-id"].unique())
+    col_names = [
+        f"Nr {number + 1} best structure" for number in range(number_of_struct)
+        ] * number_of_cluster
+    best_struct_df = best_struct_df.assign(Structure=col_names)
+    best_struct_df = best_struct_df.pivot_table(
+        index=["cluster-id"],
+        columns=["Structure"],
+        values="model",
+        aggfunc=lambda x: x,
+        )
+    best_struct_df.reset_index(inplace=True)
+    best_struct_df.rename(columns={"cluster-id": "Cluster ID"}, inplace=True)
+    return best_struct_df
+
+
+# TODO links to be fixed!
+def _add_links(best_struct_df):
+    table_df = best_struct_df.copy()
+    for col_name in table_df.columns[1:]:
+        file_name = best_struct_df[col_name].apply(lambda row: Path(row).name)
+        # href empty for local files
+        # TODO fix links to Download and Visibility
+        # we can use html href syntax but html should be responsive!
+        dl_text = "Download"
+        vis_text = "Visibility"
+        table_df[col_name] = file_name + ", " + dl_text + ", " + vis_text
+    return table_df
+
+
+def clean_capri_table(dfcl):
+    """
+    Craete a tidy capri table for the report.
+
+    It changes the clomuns names of capri table (a dataframe that is read by
+    read_capri_table). It also combines mean and std values in one column.
+
+    Parameters
+    ----------
+    dfcl : pandas DataFrame
+        dataframe of capri values
+
+    Returns
+    -------
+    dfcl : pandas DataFrame
+        DataFrame of capri table with new column names
+    """
+    dfcl = dfcl.sort_values(by=["cluster_id"])
```

Comment:
related to the previous comment: it might happen that the run gives rise to hundreds/thousands of clusters..in that case I am not sure it makes sense to list all of them and for sure they should not be ordered by cluster-id (the best cluster (lower `model-cluster-ranking`) might have `cluster_id = 2000`)

Reply:
see my comment above.